### PR TITLE
Early Adjustment Windows

### PR DESCRIPTION
**Pull request by fosskers**
The value `prereduction` found in `Chainweb.Difficulty` is used to artificially slow mining during initial conditions of the network. Its specific value is mostly a guess based on behaviour observed during testing. 

The downsides to introducing `prereduction` are multifold:

- The Genesis blocks need to be mined, since only certain `Nonce` values will allow validation to succeed.
- The Genesis blocks need to be hard-coded, since re-mining them on-the-fly isn't realistic.
- Hard-coding the blocks is boilerplate, and will only get worse with time as more `ChainwebVersion`s are invented.
- A side tool needed to be written to perform said hard-coding.
- Unit tests slow down, since the genesis blocks need to be re-mined in order to be validated.

This PR adds early adjustment windows for `BlockHeight`s lower than the first "official" adjustment.

### CODE REVIEW

**Comment by fosskers** [comment]
Rebased onto `master`.

**Comment by fosskers** [comment]
Rebased again.

**Comment by fosskers** [comment]
Rebased again.

**Comment by fosskers** [comment]
Yet another rebase.

File: src/Chainweb/TreeDB/Difficulty.hs [hunk 1]
```diff
@@ -61,14 +61,23 @@ hashTarget db bh
...
     ww :: Natural
     ww = case window ver of
-      Just (WindowWidth n) -> n
+      Just (WindowWidth n)
+          -- Intent: Initial network conditions are chaotic. `HashTarget`s are
+          -- so easy that mining progresses extremely quickly. In order to not
+          -- overwhelm the Cut network, we perform earlier adjustments before
+          -- the first "official" adjustment epoch is reached.
+          | height < n -> n `div` 10
+          | otherwise -> n
```
Comment (emilypi) [inline src/Chainweb/TreeDB/Difficulty.hs]: If we derive `Num` for `WindowWidth`, alot of this unwrapping and rewrapping goes away, ya?

File: src/Chainweb/TreeDB/Difficulty.hs [hunk 2]
```diff
@@ -61,14 +61,23 @@ hashTarget db bh
...
 
     ww :: Natural
     ww = case window ver of
-      Just (WindowWidth n) -> n
+      Just (WindowWidth n)
+          -- Intent: Initial network conditions are chaotic. `HashTarget`s are
+          -- so easy that mining progresses extremely quickly. In order to not
+          -- overwhelm the Cut network, we perform earlier adjustments before
+          -- the first "official" adjustment epoch is reached.
+          | height < n -> n `div` 10
```
Comment (emilypi) [inline src/Chainweb/TreeDB/Difficulty.hs]: Where are we deriving the div by 10 adjustment from? Not that it's wrong - I just don't know how to tell _why_ this is correct just from looking at it

File: src/Chainweb/TreeDB/Difficulty.hs (hunk 1 above)
Comment (emmanueldenloye) [reply src/Chainweb/TreeDB/Difficulty.hs]: I second this!

File: src/Chainweb/TreeDB/Difficulty.hs (hunk 2 above)
Comment (fosskers) [reply src/Chainweb/TreeDB/Difficulty.hs]: I'll add this.

File: src/Chainweb/TreeDB/Difficulty.hs (hunk 1 above)
Comment (fosskers) [reply src/Chainweb/TreeDB/Difficulty.hs]: I'll give it a shot.

**Pull request merged**